libs/geometry/src/TangentSpaceMeshInternal.h

```diff
@@ -50,57 +117,258 @@ class InternalArray {
+    }
+
+    inline size_t normalsStride() const {
+        return strideSafe<filament::math::float3>(AttributeImpl::NORMALS);
+    }
+
+    static inline size_t attributeSize(AttributeImpl attribute) {
+        switch (attribute) {
+            case AttributeImpl::UV1: return sizeof(float2);
+            case AttributeImpl::COLORS: return sizeof(float4);
+            case AttributeImpl::JOINTS: return sizeof(ushort4);
+            case AttributeImpl::WEIGHTS: return sizeof(float4);
+            case AttributeImpl::NORMALS: return sizeof(float3);
+            case AttributeImpl::UV0: return sizeof(float2);
+            case AttributeImpl::TANGENTS: return sizeof(float4);
+            case AttributeImpl::POSITIONS: return sizeof(float3);
+            case AttributeImpl::TANGENT_SPACE:
+                PANIC_POSTCONDITION("Invalid attribute found in input");                                  }
+    }
+
+    inline size_t stride(AttributeImpl attribute) const {
+        auto res = attributeData.find(attribute);
+        assert_invariant(res != attributeData.end());
+        return res->second.stride ? res->second.stride : attributeSize(attribute);
+    }
+
+    template<typename DataType>
+    inline size_t strideSafe(AttributeImpl attribute) const {
+        auto res = attributeData.find(attribute);
+        if (res == attributeData.end()) {
+            return sizeof(DataType);
+
+        }
+        return res->second.stride ? res->second.stride : sizeof(DataType);
+    }
+
+    uint8_t const* raw(AttributeImpl attribute) const {
+        switch (attribute) {
+            case AttributeImpl::UV1:
+                return (uint8_t const*) aux<float2>(attribute);
+            case AttributeImpl::COLORS:
+                return (uint8_t const*) aux<float4>(attribute);
+            case AttributeImpl::JOINTS:
+                return (uint8_t const*) aux<ushort4>(attribute);
+            case AttributeImpl::WEIGHTS:
+                return (uint8_t const*) aux<float4>(attribute);
+            case AttributeImpl::NORMALS:
+                return (uint8_t const*) aux<float3>(attribute);
+            case AttributeImpl::UV0:
+                return (uint8_t const*) aux<float2>(attribute);
+            case AttributeImpl::TANGENTS:
+                return (uint8_t const*) aux<float4>(attribute);
+            case AttributeImpl::POSITIONS:
+                return (uint8_t const*) aux<float3>(attribute);
+            case AttributeImpl::TANGENT_SPACE:
+                PANIC_POSTCONDITION("Invalid attribute found in input");
+        }
+    }
+
+    // Pass back the std::variant instead of the content.
+    InData data(AttributeImpl attribute) const {
+        auto res = attributeData.find(attribute);
+        assert_invariant(res != attributeData.end());
+        return res->second.data;
+    }
+
+    template<typename DataType>
+    inline DataType const* aux(AttributeImpl attribute) const {
```

Comment:
It was a bit confusing to me that this `aux` method can access both auxiliary data _and_ non-auxiliary data.

Reply:
Sorry, renamed to data() for consistency